RubberduckTests/Inspections/SetAssignmentWithIncompatibleObjectTypeInspectionTests.cs

```diff
@@ -0,0 +1,487 @@
+                .Object;
+
+            var inspectionResults = InspectionResults(vbe);
+
+            Assert.IsFalse(inspectionResults.Any());
+        }
+
+        [Test]
+        [Category("Inspections")]
+        public void AssignmentToOtherClassWithSameName_OneResultEach()
+        {
+            const string class1 =
+                @"Attribute VB_Exposed = True
+Private Sub Interface1_DoIt()
+End Sub
+";
+
+            const string consumerModule =
+                @"
+Private Sub TestIt()
+    Dim cls As Project2.Class1
+    Dim otherCls As Class1
+
+    Set otherCls = new Class1
+    Set cls = otherCls
+    Set otherCls = cls 
+End Sub
+";
+
+            var vbe = new MockVbeBuilder()
+                .ProjectBuilder("Project2", "project2path", ProjectProtection.Unprotected)
+                .AddComponent("Class1", ComponentType.ClassModule, class1)
+                .AddProjectToVbeBuilder()
+                .ProjectBuilder("Project1", ProjectProtection.Unprotected)
+                .AddComponent("Class1", ComponentType.ClassModule, class1)
+                .AddComponent("Module1", ComponentType.StandardModule, consumerModule)
+                .AddReference("Project2", "project2path", 0, 0, false, ReferenceKind.Project)
+                .AddProjectToVbeBuilder()
+                .Build()
+                .Object;
+
+            var inspectionResults = InspectionResults(vbe).ToList();
+
+            Assert.AreEqual(2,inspectionResults.Count());
+        }
+
+        [Test]
+        [Category("Inspections")]
+        public void LegalDowncastFromImplementedInterface_NoResult()
+        {
+            const string interface1 =
+                @"Public Sub DoIt()
+End Sub";
+
+            const string class1 =
+                @"Implements Interface1
+
+Private Sub Interface1_DoIt()
+End Sub
+";
+
+            const string consumerModule =
+                @"
+Private Sub TestIt()
+    Dim intrfc As Interface1
+    Dim cls As Class1
+
+    Set intrfc = new Class1
+    Set cls = intrfc
+End Sub
+";
+
+            var inspectionResults = InspectionResults(
+                ("Interface1", interface1, ComponentType.ClassModule),
+                ("Class1", class1, ComponentType.ClassModule),
+                ("Module1", consumerModule, ComponentType.StandardModule));
+
+            Assert.IsFalse(inspectionResults.Any());
+        }
+
+        [Test]
+        [Category("Inspections")]
+        //We cannot know whether a downcast is legal at compile time.
+        public void IllegalDowncastFromImplementedInterface_NoResult()
+        {
+            const string interface1 =
+                @"Public Sub DoIt()
+End Sub";
+
+            const string class1 =
+                @"Implements Interface1
+
+Private Sub Interface1_DoIt()
+End Sub
+";
+
+            const string consumerModule =
+                @"
+Private Sub TestIt()
+    Dim intrfc As Interface1
+    Dim cls As Class1
+
+    Set intrfc = new Class2
+    Set cls = intrfc
```

Comment:
I think this is best handled by a separate inspection to warn about suspicious casts, since we should be able to use code path analysis to determine that it came from `new Class2` originally. A separate inspection also means a separate severity, which would be lower since it's not as obvious that the code is flawed.

Reply:
Note that we cannot always determine all illegal downcasts. This situation is rather artifical. I would usually suspect downcasts  to be performed on objects being passed into a method through parameters whose declared type is the interface.

Should we warn about all downcasts? (Not part of this PR)

Reply:
Wouldn't that literally be the expectation in a code base that's written against interfaces? e.g. a factory method that creates a `New Something` but returns an `ISomething`?

Reply:
Why would a downcast be expected when writing against interfaces?

Given that `Something` implements `ISomething`,
```vba
Private Function Upcast(cls As Something) As ISomething
    Set Upcast = cls
End Function
```
is an upcast and
```vba
Private Function Downcast(cls As ISomething) As Something
    Set Downcast = cls
End Function
```
a downcast.